### PR TITLE
Add containsOnlyInstancesOf() to ObjectEnumerableAssert

### PR DESCRIPTION
**Pull request by pmwmedia**
Add a method to verify that all elements of an iterable or array are instances of given classes or interfaces.

Example:
```
Iterable objects = Arrays.asList(new String(), new StringBuilder());
assertThat(objects).containsOnlyInstancesOf(CharSequence.class);
```

#### Check List:
* Fixes N/A
* Unit tests : YES
* Javadoc with a code example (API only) : YES

### CODE REVIEW

**Comment by PascalSchumacher** [comment]
Thanks for the pull request! :+1: 

I think it would be nice if the failure message would also display the class of non-matching elements. What do you think?

**Comment by pmwmedia** [comment]
I agree with you and have implemented that the class of non-matching elements will be displayed if the class name is not part of the `toString()` method.

File: src/main/java/org/assertj/core/error/ShouldContainOnlyInstances.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    return new ShouldContainOnlyInstances(actual, types, dismatches);
+  }
+
+  private ShouldContainOnlyInstances(Object actual, Class<?>[] types, Iterable<?> dismatches) {
+    super("%n" +
+        "Expecting:%n" +
+        "  <%s>%n" +
+        "to contain only instances of:%n" +
+        "  <%s>%n" +
+        "elements are not instances:%n" +
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/error/ShouldContainOnlyInstances.java]: I would rephrase `elements are not instances:` to `but these elements are not:`

File: src/main/java/org/assertj/core/internal/Iterables.java [hunk 2]
```diff
@@ -846,6 +847,27 @@ public void assertContainsExactlyInAnyOrder(AssertionInfo info, Iterable<?> actu
                            shouldContainExactlyInAnyOrder(actual, values, notFound, notExpected, comparisonStrategy));
   }
 
+  public void assertContainsOnlyInstancesOf(AssertionInfo info, Iterable<?> actual, Class<?>[] types) {
+    checkIsNotNull(types);
+    assertNotNull(info, actual);
+
+    List<Object> dismatches = newArrayList();
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/internal/Iterables.java]: I would rename `dismatches` to `mismatches` it feels clearer to me

**Comment by joel-costigliola** [comment]
@pmwmedia thanks for the PR ! 
There are a few important missing tests: the ones that test the actual assertion logic in `Iterables` and `Arrays`, what should be added are tests like [Iterables_assertContains_Test](https://github.com/joel-costigliola/assertj-core/blob/d3dab0d22cae3028dd4c73a89a5f8bca1fd21a6d/src/test/java/org/assertj/core/internal/iterables/Iterables_assertContains_Test.java) but for `assertContainsOnlyInstancesOf `.

Would you have the time to add such tests ?

**Comment by joel-costigliola** [comment]
and sorry for the late feedback, too busy these days ...

**Comment by joel-costigliola** [comment]
For API consistency I'm renaming this assertion to `hasOnlyElementsOfTypeIn` as we already provide `hasOnlyElementsOfType`.

**Comment by pmwmedia** [comment]
What is about merging both to `hasOnlyElementsOfType(Class<?>... types)` or rename my new assertion to `hasOnlyElementsOfTypes`?

**Comment by joel-costigliola** [comment]
merging with `hasOnlyElementsOfType` would not work as `Type` is not plural.
I might go for `hasOnlyElementsOfTypes` indeed

**Comment by joel-costigliola** [comment]
Renamed to `hasOnlyElementsOfTypes`.

Thanks @pmwmedia !
